### PR TITLE
Add cache for location2pointerplan

### PR DESCRIPTION
**Pull request by murarugeorgec**
Solves #2710 
Add a cache mechanism for the Plan class.
1-to-1 mapping between a location and a possible pointer plan that was sent to a worker.
Should reduce the bandwidth usage when the same plan might be sent multiple times to the same worker.

### CODE REVIEW

File: pip-dep/requirements_dev.txt [hunk 1]
```diff
@@ -6,7 +6,9 @@ jupyter>=1.0.0
 pre-commit>=1.16.1
 pyopenssl>=19.0.0
 pyshark>=0.4.2.4
+pytest-asyncio
```
Comment (murarugeorgec) [inline pip-dep/requirements_dev.txt]: Q: Should we add versions to this?
I got a warning for this and for the scipy I got an error saying that I did not have it on my system.
Q2: Is this a good place for them?

File: test/message/test_plan.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def _(data):
 
 
 def test_raise_exception_when_sending_unbuilt_plan(workers):
-    me, bob = workers["me"], workers["bob"]
```
Comment (murarugeorgec) [inline test/message/test_plan.py]: Q: Removed me since it was not used

File: test/message/test_plan.py [hunk 3]
```diff
@@ -948,3 +948,33 @@ def plan_double_abs(x):
...
+
+def test_cached_plan_send(workers):
+    bob = workers["bob"]
+
+    @sy.func2plan(args_shape=[(1,)])
+    def plan_abs(data):
+        return data.abs()
+
+    plan_bob_ptr1 = plan_abs.send(bob)
+    plan_bob_ptr2 = plan_abs.send(bob)
```
Comment (LaRiffle) [inline test/message/test_plan.py]: Check also that plan_bob_ptr1 and plan_bob_ptr2 are the same object :) [maybe the id / id_at_location, etc]

Comment (LaRiffle) [inline test/message/test_plan.py]: Check also that plan_bob_ptr1 and plan_bob_ptr2 are the same object :) [maybe the id / id_at_location, etc]

File: pip-dep/requirements_dev.txt [hunk 4]
```diff
@@ -7,6 +7,7 @@ pre-commit>=1.16.1
 pyopenssl>=19.0.0
 pyshark>=0.4.2.4
 pytest>=4.5.0
+scipy
```
Comment (LaRiffle) [inline pip-dep/requirements_dev.txt]: Do we need it?

File: test/message/test_plan.py (hunk 3 above)
Comment (murarugeorgec) [reply test/message/test_plan.py]: I removed the ```==``` from line 965 and added an ```is``` (it should check the id of the two pointers). Should this be enough or I should also check the identity for the members?

File: pip-dep/requirements_dev.txt (hunk 4 above)
Comment (murarugeorgec) [reply pip-dep/requirements_dev.txt]: I think we do not need it since ```travis``` did not complain without it.

**Pull request merged**